scalafix-core/shared/src/main/scala/scalafix/internal/rule/MissingFinal.scala

```diff
@@ -0,0 +1,62 @@
+package scalafix.internal.rule
+
+import scala.meta._
+import scalafix.lint.LintCategory
+import scalafix.patch.Patch
+import scalafix.rule.RuleCtx
+import scalafix.{SemanticRule, SemanticdbIndex}
+
+final case class MissingFinal(index: SemanticdbIndex)
+    extends SemanticRule(
+      index,
+      "MissingFinal"
+    ) {
+
+  override def description: String =
+    "Rule that checks for or adds final modifier in corresponding places"
+
+  private lazy val error: LintCategory =
+    LintCategory.error(
+      "Some constructions should have final modifier"
+    )
+
+  override def fix(ctx: RuleCtx): Patch = {
+    def isSealed(tpe: Type): Boolean =
+      ctx.index.denotation(tpe).exists(_.isSealed)
+
+    def leaksSealedParent(mods: List[Mod], templ: Template): Boolean =
+      templ.inits.exists(i => isSealed(i.tpe)) &&
+        mods.forall(m => !m.is[Mod.Final] && !m.is[Mod.Sealed])
+
+    def addFinal(mods: Seq[Mod], d: Defn): Patch =
```

Comment:
Handy! Might might fit well in the core scalafix. I've written code like this myself a few times.